### ISSUE TITLE
etterlog: NULL pointer dereference in fingerprint_search (ec_fingerprint.c)

### ISSUE THREAD

**Issue by asarubbo** [Ettercap/ettercap#744]
There is a NULL pointer access when etterlog parses malformed file.
This is reproducible with the latest release 0.8.2 and the latest git. I'm attaching the stacktrace of the lastest git and the reproducer as zip because of the github's limitation.
[1.crash.zip](https://github.com/Ettercap/ettercap/files/412766/1.crash.zip)

```
# etterlog $FILE
Log file version    : 0.8.2                                                                                                                                                                    
Timestamp           : Thu Jul 16 15:28:54 2015 [688192]                                                                                                                                        
Type                : LOG_INFO                                                                                                                                                                 

1766 tcp OS fingerprint                                                                                                                                                                        

20530 mac vendor fingerprint                                                                                                                                                                   

2182 known services                                                                                                                                                                            

ASAN:DEADLYSIGNAL                                                                                                                                                                              
=================================================================                                                                                                                              
==9987==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7f4671a428b2 bp 0x7ffd1cdbf5b0 sp 0x7ffd1cdbf540 T0)                                                             
    #0 0x7f4671a428b1 in fingerprint_search /tmp/portage/net-analyzer/ettercap-9999/work/ettercap-9999/src/ec_fingerprint.c:189:17                                                             
    #1 0x7f4671a6ee4e in print_host /tmp/portage/net-analyzer/ettercap-9999/work/ettercap-9999/src/ec_passive.c:120:8                                                                          
    #2 0x4fe769 in display_info /tmp/portage/net-analyzer/ettercap-9999/work/ettercap-9999/utils/etterlog/el_display.c:277:10                                                                  
    #3 0x4fe769 in display /tmp/portage/net-analyzer/ettercap-9999/work/ettercap-9999/utils/etterlog/el_display.c:52                                                                           
    #4 0x507818 in main /tmp/portage/net-analyzer/ettercap-9999/work/ettercap-9999/utils/etterlog/el_main.c:94:4                                                                               
    #5 0x7f46706e561f in __libc_start_main /var/tmp/portage/sys-libs/glibc-2.22-r4/work/glibc-2.22/csu/libc-start.c:289                                                                        
    #6 0x41a408 in _start (/usr/bin/etterlog+0x41a408)                                                                                                                                         

AddressSanitizer can not provide additional info.                                                                                                                                              
SUMMARY: AddressSanitizer: SEGV /tmp/portage/net-analyzer/ettercap-9999/work/ettercap-9999/src/ec_fingerprint.c:189:17 in fingerprint_search                                                   
==9987==ABORTING                                                                                                                                                                               

etterlog 0.8.2 copyright 2001-2015 Ettercap Development Team                                                                                                                                   



==================================================
 IP address   : 192.168.0.31 

 MAC address  : 34:17:EB:9B:21:AD 
 MANUFACTURER : Dell Inc 

 DISTANCE     : 0   
 TYPE         : LAN host

 FINGERPRINT      : �
```

**Comment by koeppea** [comment]
[code]
"/home/koeppea/tmp/1.crash" is not a core dump: File format not recognized
[/code]
I'm afraid w/o **$FILE**, we're stuck here. Could you provide it please?

Moreover, in our backtrace output, I read paths like `/tmp/portage/net-analyzer/ettercap-9999/`. So I assume you've run this on a Gentoo system? But to me it looks like ettercap has been installed via **emerge** rather than directly compiled from github?

**Comment by koeppea** [comment]
> his on a Gentoo system

Or was it a *BSD?

**Comment by koeppea** [comment]
Sorry I see, `1.crash` is $FILE in this case.
I see the problem. You've inserted a null-byte in the IP header. Removing it, correctly processes the logfile. I think the issue is common over the complete bunch of issues you've uncovered. Lets see ...

**Comment by asarubbo** [comment]
Hello,
yes, it is a gentoo system and ettercap is installed via package manager. But our package manager, if an ebuild -9999 is present, is able to fetch the sources via git and compile them.
So, in other words when you install an ebuild-9999 you have a synced version of the master git.
If you need something else feel free to ask.

**Comment by LocutusOfBorg** [comment]
I merged the patch, can anybody please confirm the patch works?

**Comment by koeppea** [comment]
merged into master directly?

**Comment by LocutusOfBorg** [comment]
yep https://github.com/Ettercap/ettercap/pull/758/files

**Comment by koeppea** [comment]
Unfortunately not:
```
(gdb) r 1.crash
Starting program: /home/koeppea/tmp/ettercap/bin/etterlog 1.crash
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

etterlog 0.8.2 copyright 2001-2015 Ettercap Development Team

Log file version    : 0.8.2
Timestamp           : Thu Jul 16 17:28:54 2015 [688192]
Type                : LOG_INFO

1766 tcp OS fingerprint

20530 mac vendor fingerprint

2182 known services



==================================================
 IP address   : 192.168.0.31 

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7b7b33a in print_host (h=0x61cab0) at /home/koeppea/dev/ettercap/src/ec_passive.c:95
95	   if (GBL_CONF->geoip_support_enable)
(gdb) bt
#0  0x00007ffff7b7b33a in print_host (h=0x61cab0) at /home/koeppea/dev/ettercap/src/ec_passive.c:95
#1  0x0000000000404b1e in display_info ()
    at /home/koeppea/dev/ettercap/utils/etterlog/el_display.c:277
#2  0x000000000040434b in display () at /home/koeppea/dev/ettercap/utils/etterlog/el_display.c:52
#3  0x00000000004064d8 in main (argc=2, argv=0x7fffffffe2b8)
    at /home/koeppea/dev/ettercap/utils/etterlog/el_main.c:94
(gdb) 
```
Note, that I'm only able to reproduce the issue on 64-bit systems. On 32-bit systems, etterlog is even not able to read the timestamp of the log file.

**Comment by LocutusOfBorg** [comment]
The geoip stuff is something I'm working on it, it needs changes on libettercap I'm not sure I'm able to perform (part of the big pull request)
If you disable GEOIP and try again it should/might work

**Comment by koeppea** [comment]
Fixed with 37b7343
